src/tqec/sketchup/zx_graph.py

```diff
@@ -279,3 +353,109 @@ def __eq__(self, other: object) -> bool:
+                [
+                    subgraph | {edge_between_cur_parent}
+                    for subgraph in self._find_correlation_subgraphs_dfs(
+                        cur_corr_node, cur_zx_node.node_type, set(visited_positions)
+                    )
+                ]
+            )
+        if not branched_subgraphs:
+            return [set()]
+        # the color of node matches the correlation type
+        # broadcast the correlation type to all the neighbors
+        if parent_zx_node_type == parent_corr_type:
+            for prod in itertools.product(*branched_subgraphs):
+                correlation_subgraphs.append(set(itertools.chain(*prod)))
+            return correlation_subgraphs
+
+        # the color of node does not match the correlation type
+        # only one of the neighbors can be the correlation path
+        correlation_subgraphs.extend(itertools.chain(*branched_subgraphs))
+        return correlation_subgraphs
+
+    def find_correlation_subgraphs(self) -> list[ZXGraph]:
+        """Find the correlation subgraphs of the ZX graph.
+
+        Here a correlation subgraph is defined as a subgraph of the `ZXGraph`
+        that represents the correlation surface within a 3D spacetime diagram.
+        Each node in the correlation subgraph is composed of its position and
+        the correlation surface type, which is either `NodeType.X` or `NodeType.Z`.
+
+        For the closed diagram, the correlation
+        subgraph represents the correlation between the measured logical observables.
+        For the open diagram, the correlation subgraph represents the correlation
+        between the measured logical observables and the input/output observables,
+        which can be combined with the expected stabilizer flow to verify the correctness
+        of the computation.
+        """
+        single_node_correlation_subgraphs: list[ZXGraph] = []
+        multi_edges_correlation_subgraphs: dict[frozenset[ZXEdge], ZXGraph] = {}
+        num_subgraphs = 0
+        for node in self.isolated_nodes:
+            if node.is_virtual:
+                continue
+            subgraph = ZXGraph(f"Correlation subgraph {num_subgraphs}")
+            subgraph.add_node(node.position, node.node_type)
+            single_node_correlation_subgraphs.append(subgraph)
+            num_subgraphs += 1
+
+        def add_subgraphs(node: ZXNode, correlation_type: NodeType) -> None:
+            nonlocal num_subgraphs
+            root_corr_node = ZXNode(node.position, correlation_type)
+            for edges in self._find_correlation_subgraphs_dfs(
+                root_corr_node, node.node_type, set()
+            ):
+                if frozenset(edges) in multi_edges_correlation_subgraphs:
+                    continue
+                subgraph = ZXGraph(
+                    f"Correlation subgraph {num_subgraphs} of {self.name}"
+                )
+                subgraph.add_node(node.position, correlation_type)
```

Comment:
I think this line is not really needed as the parent node will appear in at least one of the edges. Am I missing something?
```suggestion
```